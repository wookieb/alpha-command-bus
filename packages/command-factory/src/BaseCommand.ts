export class BaseCommand {
    constructor(readonly command: string) {
    }

    freeze() {
        Object.freeze(this);
    }
}
