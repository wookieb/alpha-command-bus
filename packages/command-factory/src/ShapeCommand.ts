import {BaseCommand} from './BaseCommand';

export namespace ShapeCommand {
    export function create(name: string) {
        return class extends BaseCommand {
            static COMMAND_NAME = name;

            constructor() {
                super(name);
            }
        }
    }

    export interface Shape<TName extends string = string> {
        readonly COMMAND_NAME: TName,

        new(...args: any[]): any
    }

    export function isType(value: any): value is Shape {
        return value && 'COMMAND_NAME' in value;
    }

    export function isTypeOfName<TName extends string>(name: TName, value: any): value is Shape<TName> {
        return isType(value) && value.COMMAND_NAME === name;
    }
}
