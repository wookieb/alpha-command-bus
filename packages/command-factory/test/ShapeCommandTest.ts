import {ShapeCommand} from '@src/ShapeCommand';
import {BaseCommand} from '@src/BaseCommand';

describe('ShapeCommand', () => {
    const NAME = 'foooo';

    class Foo extends ShapeCommand.create(NAME) {
        constructor(readonly id: string) {
            super();
            this.freeze();
        }
    }

    it('properties', () => {
        expect(Foo.COMMAND_NAME)
            .toEqual(NAME);

        const command = new Foo('10');

        expect(command)
            .toMatchSnapshot();
    });


    describe('testing type', () => {
        class Bar extends BaseCommand {
            constructor() {
                super(NAME);
            }
        }

        class FooBar extends ShapeCommand.create('foobar') {
            constructor(readonly id: string) {
                super();
                this.freeze();
            }
        }

        it('simple', () => {
            expect(ShapeCommand.isType(Foo))
                .toEqual(true);

            expect(ShapeCommand.isType(FooBar))
                .toEqual(true);

            expect(ShapeCommand.isType(Bar))
                .toEqual(false);
        });

        it('with name', () => {
            expect(ShapeCommand.isTypeOfName(NAME, Foo))
                .toEqual(true);

            expect(ShapeCommand.isTypeOfName(NAME, FooBar))
                .toEqual(false);

            expect(ShapeCommand.isTypeOfName(NAME, Bar))
                .toEqual(false);
        });
    })
});
