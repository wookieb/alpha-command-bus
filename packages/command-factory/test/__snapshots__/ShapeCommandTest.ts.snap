// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShapeCommand properties 1`] = `
Foo {
  "command": "foooo",
  "id": "10",
}
`;
